File: robocode/robocode/peer/proxies/BasicRobotProxy.java
/*******************************************************************************
 * Copyright (c) 2001, 2008 Mathew A. Nelson and Robocode contributors
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Common Public License v1.0
 * which accompanies this distribution, and is available at
 * http://robocode.sourceforge.net/license/cpl-v10.html
 *
 * Contributors:
 *     Pavel Savara
 *     - Initial implementation
 *******************************************************************************/
package robocode.peer.proxies;


import robocode.*;
import robocode.Event;
import robocode.exception.DisabledException;
import robocode.exception.RobotException;
import robocode.manager.IHostManager;
import robocode.peer.*;
import robocode.peer.robot.EventManager;
import robocode.peer.robot.RobotClassManager;
import robocode.peer.robot.TeamMessage;
import robocode.robotinterfaces.peer.IBasicRobotPeer;
import robocode.robotpaint.Graphics2DProxy;
import robocode.util.Utils;

import java.awt.*;
import java.util.concurrent.atomic.AtomicInteger;


/**
 * @author Pavel Savara (original)
 */
public class BasicRobotProxy extends HostingRobotProxy implements IBasicRobotPeer {
	private static final long
			MAX_SET_CALL_COUNT = 10000,
			MAX_GET_CALL_COUNT = 10000;

	private Graphics2DProxy graphicsProxy;

	protected RobotStatus status;
	protected RobotCommands commands;
	private ExecResult execResult;

	private AtomicInteger setCallCount = new AtomicInteger(0);
	private AtomicInteger getCallCount = new AtomicInteger(0);

	protected Condition waitCondition;
	protected boolean testingCondition;
	protected double firedEnergy;
	protected double firedHeat;

	public BasicRobotProxy(RobotClassManager robotClassManager, IHostManager hostManager, IRobotPeer peer, RobotStatics statics) {
		super(robotClassManager, hostManager, peer, statics);

		eventManager = new EventManager(this);

		graphicsProxy = new Graphics2DProxy();

		// dummy
		execResult = new ExecResult(null, null, null, null, false, false);

		setSetCallCount(0);
		setGetCallCount(0);
	}

	protected void initializeRound(RobotCommands commands, RobotStatus status) {
		updateStatus(commands, status);
		eventManager.reset();
		final StatusEvent start = new StatusEvent(status);

		start.setTime(0);
		eventManager.add(start);
	}

	@Override
	public void cleanup() {
		super.cleanup();

		// Cleanup and remove current wait condition
		if (waitCondition != null) {
			waitCondition.cleanup();
			waitCondition = null;
		}

		// Cleanup and remove the event manager
		if (eventManager != null) {
			eventManager.cleanup();
			eventManager = null;
		}

		// Cleanup graphics proxy
		graphicsProxy = null;
		execResult = null;
		status = null;
		commands = null;
	}

	// asynchronous actions
	public Bullet setFire(double power) {
		return setFireImpl(power);
	}

	// blocking actions
	public void execute() {
		executeImpl();
	}

	public void move(double distance) {
		setMoveImpl(distance);
		do {
			execute(); // Always tick at least once
		} while (getDistanceRemaining() != 0);
	}

	public void turnBody(double radians) {
		setTurnBodyImpl(radians);
		do {
			execute(); // Always tick at least once
		} while (getBodyTurnRemaining() != 0);
	}

	public void turnGun(double radians) {
		setTurnGunImpl(radians);
		do {
			execute(); // Always tick at least once
		} while (getGunTurnRemaining() != 0);
	}

	public Bullet fire(double power) {
		Bullet bullet = setFire(power);

		execute();
		return bullet;
	}

	// fast setters
	public void setBodyColor(Color color) {
		setCall();
		commands.setBodyColor(color.getRGB());
	}

	public void setGunColor(Color color) {
		setCall();
		commands.setGunColor(color.getRGB());
	}

	public void setRadarColor(Color color) {
		setCall();
		commands.setRadarColor(color.getRGB());
	}

	public void setBulletColor(Color color) {
		setCall();
		commands.setBulletColor(color.getRGB());
	}

	public void setScanColor(Color color) {
		setCall();
		commands.setScanColor(color.getRGB());
	}

	// counters
	public void setCall() {
		final int res = setCallCount.incrementAndGet();

		if (res >= MAX_SET_CALL_COUNT) {
			println("SYSTEM: You have made " + res + " calls to setXX methods without calling execute()");
			throw new DisabledException("Too many calls to setXX methods");
		}
	}

	public void getCall() {
		final int res = getCallCount.incrementAndGet();

		if (res >= MAX_GET_CALL_COUNT) {
			println("SYSTEM: You have made " + res + " calls to getXX methods without calling execute()");
			throw new DisabledException("Too many calls to getXX methods");
		}
	}

	public double getDistanceRemaining() {
		getCall();
		return commands.getDistanceRemaining();
	}

	public double getRadarTurnRemaining() {
		getCall();
		return commands.getRadarTurnRemaining();
	}

	public double getBodyTurnRemaining() {
		getCall();
		return commands.getBodyTurnRemaining();
	}

	public double getGunTurnRemaining() {
		getCall();
		return commands.getGunTurnRemaining();
	}

	public double getVelocity() {
		getCall();
		return status.getVelocity();
	}

	public double getGunCoolingRate() {
		getCall();
		return statics.getBattleRules().getGunCoolingRate();
	}

	public String getName() {
		getCall();
		return statics.getName();
	}

	public long getTime() {
		getCall();
		return getTimeImpl();
	}

	public double getBodyHeading() {
		getCall();
		return status.getHeadingRadians();
	}

	public double getGunHeading() {
		getCall();
		return status.getGunHeadingRadians();
	}

	public double getRadarHeading() {
		getCall();
		return status.getRadarHeadingRadians();
	}

	public double getEnergy() {
		getCall();
		return getEnergyImpl();
	}

	public double getGunHeat() {
		getCall();
		return getGunHeatImpl();
	}

	public double getX() {
		getCall();
		return status.getX();
	}

	public double getY() {
		getCall();
		return status.getY();
	}

	public int getOthers() {
		getCall();
		return status.getOthers();
	}

	public double getBattleFieldHeight() {
		getCall();
		return statics.getBattleRules().getBattlefieldHeight();
	}

	public double getBattleFieldWidth() {
		getCall();
		return statics.getBattleRules().getBattlefieldWidth();
	}

	public int getNumRounds() {
		getCall();
		return statics.getBattleRules().getNumRounds();
	}

	public int getRoundNum() {
		getCall();
		return status.getRoundNum();
	}

	public Graphics2D getGraphics() {
		getCall();
		return getGraphicsImpl();
	}

	// -----------
	// implementations
	// -----------

	public long getTimeImpl() {
		return status.getTime();
	}

	public Graphics2D getGraphicsImpl() {
		return graphicsProxy;
	}

	@Override
	protected final void executeImpl() {
		// Entering tick
		robotThreadManager.checkRunThread();
		if (testingCondition) {
			throw new RobotException(
					"You cannot take action inside Condition.test().  You should handle onCustomEvent instead.");
		}

		setSetCallCount(0);
		setGetCallCount(0);

		// This stops autoscan from scanning...
		if (waitCondition != null && waitCondition.test()) {
			waitCondition = null;
			commands.setScan(true);
		}

		commands.setOutputText(out.readAndReset());
		commands.setGraphicsCalls(graphicsProxy.getQueuedCalls());
		graphicsProxy.clearQueue();

		// call server
		execResult = peer.executeImpl(commands);

		updateStatus(execResult.commands, execResult.status);
		firedEnergy = 0;
		firedHeat = 0;

		// add new events first
		if (execResult.events != null) {
			for (Event event : execResult.events) {
				eventManager.add(event);
			}
		}

		// add new team messages
		loadTeamMessages(execResult.teamMessages);

		eventManager.processEvents();
	}

	@Override
	protected final void waitForBattleEndImpl() {
		eventManager.clearAllEvents(false);
		do {
			commands.setOutputText(out.readAndReset());
			commands.setGraphicsCalls(graphicsProxy.getQueuedCalls());
			graphicsProxy.clearQueue();

			// call server
			execResult = peer.waitForBattleEndImpl(commands);

			updateStatus(execResult.commands, execResult.status);

			// add new events
			if (execResult.events != null) {
				for (Event event : execResult.events) {
					if (event instanceof BattleEndedEvent) {
						eventManager.add(event);
					}
				}
			}
			eventManager.resetCustomEvents();
			eventManager.processEvents();
		} while (!execResult.halt && execResult.shouldWait);
	}

	private void updateStatus(RobotCommands commands, RobotStatus status) {
		this.status = status;
		this.commands = commands;
	}

	protected void loadTeamMessages(java.util.List<TeamMessage> teamMessages) {}

	protected final double getEnergyImpl() {
		return status.getEnergy() - firedEnergy;
	}

	protected final double getGunHeatImpl() {
		return status.getGunHeat() - firedHeat;
	}

	protected final void setMoveImpl(double distance) {
		if (getEnergyImpl() == 0) {
			return;
		}
		commands.setDistanceRemaining(distance);
		commands.setMoved(true);
	}

	protected final Bullet setFireImpl(double power) {
		if (Double.isNaN(power)) {
			println("SYSTEM: You cannot call fire(NaN)");
			return null;
		}
		if (getGunHeatImpl() > 0 || getEnergyImpl() == 0) {
			return null;
		}

		Bullet bullet;
		BulletCommand wrapper;
		Event currentTopEvent = eventManager.getCurrentTopEvent();

		if (currentTopEvent != null && currentTopEvent.getTime() == status.getTime() && !statics.isAdvancedRobot()
				&& status.getGunHeadingRadians() == status.getRadarHeadingRadians()
				&& ScannedRobotEvent.class.isAssignableFrom(currentTopEvent.getClass())) {
			// this is angle assisted bullet
			ScannedRobotEvent e = (ScannedRobotEvent) currentTopEvent;
			double fireAssistAngle = Utils.normalAbsoluteAngle(status.getHeadingRadians() + e.getBearingRadians());

			bullet = new Bullet(fireAssistAngle, getX(), getY(), power, statics.getName());
			wrapper = new BulletCommand(bullet, true, fireAssistAngle);
		} else {
			// this is normal bullet
			bullet = new Bullet(status.getGunHeadingRadians(), getX(), getY(), power, statics.getName());
			wrapper = new BulletCommand(bullet, false, 0);
		}

		firedEnergy += power;
		firedHeat += Rules.getGunHeat(power);

		commands.getBullets().add(wrapper);

		return bullet;
	}

	protected final void setTurnGunImpl(double radians) {
		commands.setGunTurnRemaining(radians);
	}

	protected final void setTurnBodyImpl(double radians) {
		if (getEnergyImpl() > 0) {
			commands.setBodyTurnRemaining(radians);
		}
	}

	protected final void setTurnRadarImpl(double radians) {
		commands.setRadarTurnRemaining(radians);
	}

	// -----------
	// battle driven methods
	// -----------

	private void setSetCallCount(int setCallCount) {
		this.setCallCount.set(setCallCount);
	}

	private void setGetCallCount(int getCallCount) {
		this.getCallCount.set(getCallCount);
	}

	// -----------
	// for robot thread
	// -----------

	public void setTestingCondition(boolean testingCondition) {
		this.testingCondition = testingCondition;
	}

	@Override
	public String toString() {
		return statics.getShortName() + "(" + (int) status.getEnergy() + ") X" + (int) status.getX() + " Y"
				+ (int) status.getY();
	}
}

File: plugins/dotnet/robocode.dotnet.nhost/src/host/proxies/BasicRobotProxy.cs
/**
 * Copyright (c) 2001-2019 Mathew A. Nelson and Robocode contributors
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * https://robocode.sourceforge.io/license/epl-v10.html
 */
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Threading;
using net.sf.jni4net.nio;
using net.sf.robocode.dotnet.host.events;
using net.sf.robocode.dotnet.peer;
using net.sf.robocode.host;
using net.sf.robocode.io;
using net.sf.robocode.peer;
using net.sf.robocode.repository;
using net.sf.robocode.security;
using net.sf.robocode.serialization;
using Robocode;
using robocode.exception;
using Robocode.Exception;
using Robocode.RobotInterfaces.Peer;
using Robocode.Util;
using ByteBuffer = net.sf.robocode.nio.ByteBuffer;
using Thread = System.Threading.Thread;

namespace net.sf.robocode.dotnet.host.proxies
{
    public class BasicRobotProxy : HostingRobotProxy, IBasicRobotPeer
    {
        private const long
            MAX_SET_CALL_COUNT = 10000,
            MAX_GET_CALL_COUNT = 10000;

        private GraphicsProxy graphicsProxy;

        private RobotStatus status;
        private bool isDisabled;
        protected ExecCommands commands;
        private ExecResults execResults;

        private readonly Dictionary<int, Bullet> bullets = new Dictionary<int, Bullet>();
        private int bulletCounter;

        private int setCallCount;
        private int getCallCount;

        protected Condition waitCondition;
        private bool testingCondition;
        private double firedEnergy;
        private double firedHeat;

        private readonly DirectByteBuffer execJavaBuffer;
        private readonly ByteBuffer execNetBuffer;
        private readonly RbSerializerN rbSerializerN;

        public BasicRobotProxy(IRobotItem specification, IHostManager hostManager, IRobotPeer peer, RobotStatics statics)
            : base(specification, hostManager, peer, statics)
        {
            eventManager = new EventManager(this);

            graphicsProxy = new GraphicsProxy();

            // dummy
            execResults = new ExecResults(null, null, null, null, null, false, false, false);

            setSetCallCount(0);
            setGetCallCount(0);

            var sharedBuffer = new byte[10*1024*100];
            execJavaBuffer = new DirectByteBuffer(sharedBuffer);
            execNetBuffer = ByteBuffer.wrap(sharedBuffer);
            rbSerializerN = new RbSerializerN();
            this.peer.setupBuffer(execJavaBuffer);
        }

        // asynchronous actions

        #region IBasicRobotPeer Members

        public Bullet SetFire(double power)
        {
            SetCall();
            return setFireImpl(power);
        }

        // blocking actions
        public void Execute()
        {
            executeImpl();
        }

        public void Move(double distance)
        {
            setMoveImpl(distance);
            do
            {
                Execute(); // Always tick at least once
            } while (GetDistanceRemaining() != 0);
        }

        public void TurnBody(double radians)
        {
            setTurnBodyImpl(radians);
            do
            {
                Execute(); // Always tick at least once
            } while (GetBodyTurnRemaining() != 0);
        }

        public void TurnGun(double radians)
        {
            setTurnGunImpl(radians);
            do
            {
                Execute(); // Always tick at least once
            } while (GetGunTurnRemaining() != 0);
        }

        public Bullet Fire(double power)
        {
            Bullet bullet = SetFire(power);

            Execute();
            return bullet;
        }

        // fast setters
        public void SetBodyColor(Color color)
        {
            SetCall();
            commands.setBodyColor(color.ToArgb());
        }

        public Color GetBodyColor()
        {
            GetCall();
            return Color.FromArgb(commands.getBodyColor());
        }

        public void SetGunColor(Color color)
        {
            SetCall();
            commands.setGunColor(color.ToArgb());
        }

        public Color GetGunColor()
        {
            GetCall();
            return Color.FromArgb(commands.getGunColor());
        }

        public void SetRadarColor(Color color)
        {
            SetCall();
            commands.setRadarColor(color.ToArgb());
        }

        public Color GetRadarColor()
        {
            GetCall();
            return Color.FromArgb(commands.getRadarColor());
        }

        public void SetBulletColor(Color color)
        {
            SetCall();
            commands.setBulletColor(color.ToArgb());
        }

        public Color GetBulletColor()
        {
            GetCall();
            return Color.FromArgb(commands.getBulletColor());
        }

        public void SetScanColor(Color color)
        {
            SetCall();
            commands.setScanColor(color.ToArgb());
        }

        public Color GetScanColor()
        {
            GetCall();
            return Color.FromArgb(commands.getScanColor());
        }

        // counters
        public void SetCall()
        {
            if (!isDisabled)
            {
                int res = Interlocked.Increment(ref setCallCount);

                if (res >= MAX_SET_CALL_COUNT)
                {
                    isDisabled = true;
                    println("SYSTEM: You have made " + res + " calls to setXX methods without calling Execute()");
                    throw new DisabledException("Too many calls to setXX methods");
                }
            }
        }

        public void GetCall()
        {
            if (!isDisabled)
            {
                int res = Interlocked.Increment(ref getCallCount);

                if (res >= MAX_GET_CALL_COUNT)
                {
                    isDisabled = true;
                    println("SYSTEM: You have made " + res + " calls to getXX methods without calling Execute()");
                    throw new DisabledException("Too many calls to getXX methods");
                }
            }
        }

        public double GetDistanceRemaining()
        {
            GetCall();
            return commands.getDistanceRemaining();
        }

        public double GetRadarTurnRemaining()
        {
            GetCall();
            return commands.getRadarTurnRemaining();
        }

        public double GetBodyTurnRemaining()
        {
            GetCall();
            return commands.getBodyTurnRemaining();
        }

        public double GetGunTurnRemaining()
        {
            GetCall();
            return commands.getGunTurnRemaining();
        }

        public double GetVelocity()
        {
            GetCall();
            return status.Velocity;
        }

        public double GetGunCoolingRate()
        {
            GetCall();
            return statics.getBattleRules().GunCoolingRate;
        }

        public String GetName()
        {
            GetCall();
            return statics.getName();
        }

        public long GetTime()
        {
            GetCall();
            return getTimeImpl();
        }

        public double GetBodyHeading()
        {
            GetCall();
            return status.HeadingRadians;
        }

        public double GetGunHeading()
        {
            GetCall();
            return status.GunHeadingRadians;
        }

        public double GetRadarHeading()
        {
            GetCall();
            return status.RadarHeadingRadians;
        }

        public double GetEnergy()
        {
            GetCall();
            return getEnergyImpl();
        }

        public double GetGunHeat()
        {
            GetCall();
            return getGunHeatImpl();
        }

        public double GetX()
        {
            GetCall();
            return status.X;
        }

        public double GetY()
        {
            GetCall();
            return status.Y;
        }

        public int GetOthers()
        {
            GetCall();
            return status.Others;
        }

        public int GetNumSentries()
        {
            GetCall();
            return status.NumSentries;
        }

        public double GetBattleFieldHeight()
        {
            GetCall();
            return statics.getBattleRules().BattlefieldHeight;
        }

        public double GetBattleFieldWidth()
        {
            GetCall();
            return statics.getBattleRules().BattlefieldWidth;
        }

        public int GetNumRounds()
        {
            GetCall();
            return statics.getBattleRules().NumRounds;
        }

        public int GetRoundNum()
        {
            GetCall();
            return status.RoundNum;
        }

        public int GetSentryBorderSize()
        {
            GetCall();
            return statics.getBattleRules().SentryBorderSize;
        }

        public IGraphics GetGraphics()
        {
            GetCall();
            commands.setTryingToPaint(true);
            return getGraphicsImpl();
        }

        public void SetDebugProperty(String key, String value)
        {
            SetCall();
            commands.setDebugProperty(key, value);
        }

        public void Rescan()
        {
            bool reset = false;
            bool origInterruptableValue = false;

            if (eventManager.CurrentTopEventPriority == eventManager.ScannedRobotEventPriority)
            {
                reset = true;
                origInterruptableValue = eventManager.IsInterruptible(eventManager.ScannedRobotEventPriority);
                eventManager.SetInterruptible(eventManager.ScannedRobotEventPriority, true);
            }

            commands.setScan(true);
            executeImpl();

            if (reset)
            {
                eventManager.SetInterruptible(eventManager.ScannedRobotEventPriority, origInterruptableValue);
            }
        }

        #endregion

        internal override void initializeRound(ExecCommands commands, RobotStatus status)
        {
            updateStatus(commands, status);
            eventManager.Reset();
            var start = new StatusEvent(status);

            eventManager.Add(start);
            setSetCallCount(0);
            setGetCallCount(0);
        }

        public override void cleanup()
        {
            base.cleanup();

            // Cleanup and remove current wait condition
            waitCondition = null;

            // Cleanup and remove the event manager
            if (eventManager != null)
            {
                eventManager.Cleanup();
                eventManager = null;
            }

            graphicsProxy = null;
            execResults = null;
            status = null;
            commands = null;
        }

        // -----------
        // implementations
        // -----------

        public long getTimeImpl()
        {
            return status.Time;
        }

        public IGraphics getGraphicsImpl()
        {
            return graphicsProxy;
        }

        protected override sealed void executeImpl()
        {
            if (execResults == null)
            {
                // this is to slow down undead robot after cleanup, from fast exception-loop
                Thread.Sleep(1000);
            }

            // Entering tick
            if (testingCondition)
            {
                throw new RobotException(
                    "You cannot take action inside Condition.Test().  You should handle OnCustomEvent instead.");
            }

            setSetCallCount(0);
            setGetCallCount(0);

            // This stops autoscan from scanning...
            if (waitCondition != null && waitCondition.Test())
            {
                waitCondition = null;
                commands.setScan(true);
            }

            commands.setOutputText(GetOutTextAndReset());
            commands.setGraphicsCalls(graphicsProxy.readoutQueuedCalls());

            // call server
            SerializeCommands();
            peer.executeImplSerial();
            DeserializeResults();

            if (execResults == null)
            {
                throw new InvalidOperationException();
            }

            updateStatus(execResults.getCommands(), execResults.getStatus());

            graphicsProxy.setPaintingEnabled(execResults.isPaintEnabled());
            firedEnergy = 0;
            firedHeat = 0;

            // add new events
            eventManager.Add(new StatusEvent(execResults.getStatus()));
            if (statics.IsPaintRobot() && execResults.isPaintEnabled())
            {
                // Add paint event, if robot is a paint robot and its painting is enabled
                eventManager.Add(new PaintEvent());
            }

            // add other events
            if (execResults.getEvents() != null)
            {
                foreach (Event evnt in execResults.getEvents())
                {
                    eventManager.Add(evnt);
                    HiddenAccessN.UpdateBullets(evnt, bullets);
                }
            }

            if (execResults.getBulletUpdates() != null)
            {
                foreach (BulletStatus bulletStatus in execResults.getBulletUpdates())
                {
                    Bullet bullet;
                    if (bullets.TryGetValue(bulletStatus.bulletId, out bullet))
                    {
                        HiddenAccessN.Update(bullet, bulletStatus.x, bulletStatus.y, bulletStatus.victimName, bulletStatus.isActive);
                        if (!bulletStatus.isActive)
                        {
                            bullets.Remove(bulletStatus.bulletId);
                        }
                    }
                }
            }

            // add new team messages
            loadTeamMessages(execResults.getTeamMessages());

            eventManager.ProcessEvents();
        }

        private string GetOutTextAndReset()
        {
            string res;
            lock (output)
            {
                output.Flush();
                res = outputSb.ToString();
                outputSb.Length = 0;
            }
            return res;
        }


        private void SerializeCommands()
        {
            execNetBuffer.clear();
            rbSerializerN.serializeToBuffer(execNetBuffer, RbSerializerN.ExecCommands_TYPE, commands);
            execJavaBuffer.Position(0);
            execJavaBuffer.Limit(execNetBuffer.limit());
        }

        private void DeserializeResults()
        {
            execNetBuffer.position(0);
            execNetBuffer.limit(execJavaBuffer.limit());
            execResults = (ExecResults) rbSerializerN.deserialize(execNetBuffer);
        }

        protected override sealed void waitForBattleEndImpl()
        {
            eventManager.ClearAllEvents(false);
            graphicsProxy.setPaintingEnabled(false);
            do
            {
                commands.setOutputText(GetOutTextAndReset());
                commands.setGraphicsCalls(graphicsProxy.readoutQueuedCalls());

                // call server
                try
                {
                    SerializeCommands();
                    peer.waitForBattleEndImplSerial();
                    DeserializeResults();
                }
                catch (Exception ex)
                {
                    LoggerN.logMessage(statics.getName() + ": Exception: " + ex); // without stack here
                    return;
                }

                updateStatus(execResults.getCommands(), execResults.getStatus());

                // add new events
                if (execResults.getEvents() != null)
                {
                    foreach (Event evnt in execResults.getEvents())
                    {
                        if (evnt is BattleEndedEvent)
                        {
                            eventManager.Add(evnt);
                        }
                    }
                }
                eventManager.ResetCustomEvents();

                eventManager.ProcessEvents();

            } while (!execResults.isHalt() && execResults.isShouldWait());
        }

        private void updateStatus(ExecCommands commands, RobotStatus status)
        {
            this.status = status;
            this.commands = commands;
        }

        protected virtual void loadTeamMessages(List<TeamMessage> teamMessages)
        {
        }

        protected double getEnergyImpl()
        {
            return status.Energy - firedEnergy;
        }

        protected double getGunHeatImpl()
        {
            return status.GunHeat + firedHeat;
        }

        protected void setMoveImpl(double distance)
        {
            if (getEnergyImpl() == 0)
            {
                return;
            }
            commands.setDistanceRemaining(distance);
            commands.setMoved(true);
        }

        protected Bullet setFireImpl(double power)
        {
            if (Double.IsNaN(power))
            {
                println("SYSTEM: You cannot call Fire(NaN)");
                return null;
            }
            if (getGunHeatImpl() > 0 || getEnergyImpl() == 0)
            {
                return null;
            }

            power = Math.Min(getEnergyImpl(), Math.Min(Math.Max(power, Rules.MIN_BULLET_POWER), Rules.MAX_BULLET_POWER));

            Bullet bullet;
            BulletCommand wrapper;
            Event currentTopEvent = eventManager.CurrentTopEvent;

            bulletCounter++;

            if (currentTopEvent != null && currentTopEvent.Time == status.Time && !statics.IsAdvancedRobot()
                && status.GunHeadingRadians == status.RadarHeadingRadians
                && typeof (ScannedRobotEvent).IsAssignableFrom(currentTopEvent.GetType()))
            {
                // this is angle assisted bullet
                var e = (ScannedRobotEvent) currentTopEvent;
                double fireAssistAngle = Utils.NormalAbsoluteAngle(status.HeadingRadians + e.BearingRadians);

                bullet = new Bullet(fireAssistAngle, GetX(), GetY(), power, statics.getName(), null, true, bulletCounter);
                wrapper = new BulletCommand(power, true, fireAssistAngle, bulletCounter);
            }
            else
            {
                // this is normal bullet
                bullet = new Bullet(status.GunHeadingRadians, GetX(), GetY(), power, statics.getName(), null, true,
                                    bulletCounter);
                wrapper = new BulletCommand(power, false, 0, bulletCounter);
            }

            firedEnergy += power;
            firedHeat += Rules.GetGunHeat(power);

            commands.getBullets().Add(wrapper);

            bullets.Add(bulletCounter, bullet);

            return bullet;
        }

        protected void setTurnGunImpl(double radians)
        {
            commands.setGunTurnRemaining(radians);
        }

        protected void setTurnBodyImpl(double radians)
        {
            if (getEnergyImpl() > 0)
            {
                commands.setBodyTurnRemaining(radians);
            }
        }

        protected void setTurnRadarImpl(double radians)
        {
            commands.setRadarTurnRemaining(radians);
        }

        // -----------
        // battle driven methods
        // -----------

        private void setSetCallCount(int setCallCount)
        {
            this.setCallCount = setCallCount;
        }

        private void setGetCallCount(int getCallCount)
        {
            this.getCallCount = getCallCount;
        }

        // -----------
        // for robot thread
        // -----------

        public void setTestingCondition(bool testingCondition)
        {
            this.testingCondition = testingCondition;
        }


        public override String ToString()
        {
            return statics.getShortName() + "(" + (int) status.Energy + ") X" + (int) status.X + " Y"
                   + (int) status.Y;
        }
    }
}
